comment/service/email.py

```diff
@@ -0,0 +1,82 @@
+        context['contact'] = settings.COMMENT_CONTACT_EMAIL
+        return context
+
+    def get_message(self, subject, body, receivers, html_msg=None):
+        msg = EmailMultiAlternatives(subject, body, self.sender, receivers)
+        if html_msg:
+            msg.attach_alternative(html_msg, 'text/html')
+        return msg
+
+    def send_messages(self, messages):
+        connection = get_connection()  # Use default email connection
+        self.thread = Thread(target=connection.send_messages, args=(messages,))
+        self.thread.start()
+
+    def get_message_templates(self, text_template, html_template, msg_context):
+        text_msg_template = loader.get_template(text_template)
+        text_msg = text_msg_template.render(msg_context)
+        html_msg = None
+        if self.is_html:
+            html_msg_template = loader.get_template(html_template)
+            html_msg = html_msg_template.render(msg_context)
+        return text_msg, html_msg
+
+    def send_confirmation_request(self, api=False):
+        comment_dict = self.comment.to_dict()
+        receivers = [comment_dict['email']]
+        key = signing.dumps(comment_dict, compress=True)
+        text_template = 'comment/anonymous/confirmation_request.txt'
+        html_template = 'comment/anonymous/confirmation_request.html'
+        subject = EmailInfo.CONFIRMATION_SUBJECT
+        if api:
+            confirmation_url = f'/api/comments/confirm/{key}/'
+        else:
+            confirmation_url = reverse('comment:confirm-comment', args=[key])
+
+        context = self.get_msg_context(confirmation_url=confirmation_url)
+        text_msg, html_msg = self.get_message_templates(text_template, html_template, context)
+        msg = self.get_message(subject, text_msg, receivers, html_msg=html_msg)
+        self.send_messages([msg])
+
+    def send_notification_to_followers(self):
+        thread = self.comment.content_object if self.comment.is_parent else self.comment.parent
+        thread_name = str(self.comment.content_object) \
+            if self.comment.is_parent else str(self.comment.parent).split(':')[0]
+        followers = Follower.objects.filter_for_model_object(thread).exclude(email=self.comment.email)
+        if not followers:
+            return
+        username = get_username_for_comment(self.comment)
+        subject = EmailInfo.NOTIFICATION_SUBJECT.format(username=username, thread_name=thread_name)
+        messages = []
+        text_template = 'comment/notifications/notification.txt'
+        html_template = 'comment/notifications/notification.html'
+        for follower in followers:
+            context = self.get_msg_context(thread_name=thread_name, receiver=follower.username)
+            text_msg, html_msg = self.get_message_templates(text_template, html_template, context)
+            messages.append(self.get_message(subject, text_msg, [follower.email], html_msg=html_msg))
+        self.send_messages(messages)
```

Comment:
We can probably transform this function to something to along these lines to make this more readable and also separate concerns.

```python
   def get_thread(self):
       if self.comment.is_parent:
           return self.comment.content_object
       return self.comment.parent

   def get_thread_name(self):
       if self.comment.is_parent:
           return str(self.comment.content_object)
       return str(self.comment.parent).split(':')[0]
   
   def get_subject(self, thread_name):
       username = get_username_for_comment(self.comment)
       return EmailInfo.NOTIFICATION_SUBJECT.format(username=username, thread_name=thread_name)
        
   def get_messages(thread_name, receivers):
       text_template = 'comment/notifications/notification.txt'
       html_template = 'comment/notifications/notification.html'
       subject = self.get_subject(thread_name)
       messages = []
       for receiver in receivers:
          context = self.get_msg_context(thread_name=thread_name, receiver=receiver.username)
          text_msg, html_msg = self.get_message_templates(text_template, html_template, context)
          messages.append(self.get_message(subject, text_msg, [receiver.email], html_msg=html_msg))
        return  messages
        
    def send_notification_to_followers(self):
       thread = self.get_thread()
       followers = Follower.objects.filter_for_model_object(thread).exclude(email=self.comment.email)
        if not followers:
            return
        thread_name = self.get_thread_name()
        messages = self.get_messages(thread_name, followers)
        self.send_messages(messages)
```

Reply:
nice suggestion!